File: services/ai-model/src/ai_model/domain/agent_config.py
"""Agent configuration domain models for AI Model service.

This module defines the Pydantic models for agent configuration storage.
Agent configs are stored in the ai_model.agent_configs MongoDB collection.

Key design decisions (see architecture/ai-model-architecture/key-decisions.md):
- Single collection with agent_type discriminator field
- Pydantic discriminated unions handle 5 agent types
- Version + status lifecycle for config management

Agent Types:
- extractor: Structured data extraction from unstructured input
- explorer: Multi-step analysis with conditional routing (RAG-enabled)
- generator: Content generation (plans, reports, messages) (RAG-enabled)
- conversational: Multi-turn dialogue with context management (RAG-enabled)
- tiered-vision: Cost-optimized image analysis with two-tier processing
"""

from datetime import UTC, datetime
from enum import Enum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

# =============================================================================
# ENUMS
# =============================================================================


class AgentType(str, Enum):
    """Agent type enum.

    Defines the 5 agent types supported by the AI Model service.
    Each type has distinct configuration requirements and execution patterns.
    """

    EXTRACTOR = "extractor"
    EXPLORER = "explorer"
    GENERATOR = "generator"
    CONVERSATIONAL = "conversational"
    TIERED_VISION = "tiered-vision"


class AgentConfigStatus(str, Enum):
    """Agent configuration lifecycle status.

    Status transitions:
    - draft: Development, validation relaxed
    - staged: Ready for promotion
    - active: Currently in use (only one per agent_id)
    - archived: Historical version, kept for audit

    Transition flow: draft -> staged -> active -> archived
    """

    DRAFT = "draft"
    STAGED = "staged"
    ACTIVE = "active"
    ARCHIVED = "archived"


# =============================================================================
# SHARED COMPONENT MODELS
# =============================================================================


class LLMConfig(BaseModel):
    """LLM configuration for agent execution.

    Specifies the model and parameters for LLM calls.
    Model is explicit per agent (no task_type routing).
    """

    model: str = Field(description="Explicit model identifier, e.g. 'anthropic/claude-3-5-sonnet'")
    temperature: float = Field(
        default=0.3,
        ge=0.0,
        le=2.0,
        description="Sampling temperature (0.0-2.0)",
    )
    max_tokens: int = Field(
        default=2000,
        ge=1,
        le=100000,
        description="Maximum tokens in response",
    )


class RAGConfig(BaseModel):
    """RAG retrieval configuration.

    Configures knowledge retrieval for RAG-enabled agents
    (Explorer, Generator, Conversational, Tiered-Vision Tier 2).
    """

    enabled: bool = Field(
        default=True,
        description="Whether RAG is enabled for this agent",
    )
    query_template: str | None = Field(
        default=None,
        description="Template for RAG query construction with {{variables}}",
    )
    knowledge_domains: list[str] = Field(
        default_factory=list,
        description="Knowledge domains to search (e.g., 'plant_diseases', 'tea_cultivation')",
    )
    top_k: int = Field(
        default=5,
        ge=1,
        le=100,
        description="Number of documents to retrieve",
    )
    min_similarity: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Minimum similarity threshold for results",
    )


class InputConfig(BaseModel):
    """Agent input contract.

    Defines the event that triggers the agent and expected input schema.
    """

    event: str = Field(description="DAPR event topic that triggers this agent (e.g., 'collection.document.received')")
    schema: dict[str, Any] = Field(description="JSON schema defining required and optional input fields")


class OutputConfig(BaseModel):
    """Agent output contract.

    Defines the event published on completion and output schema.
    """

    event: str = Field(description="DAPR event topic for completion (e.g., 'ai.extraction.complete')")
    schema: dict[str, Any] = Field(description="JSON schema defining output fields")


class MCPSourceConfig(BaseModel):
    """MCP server data source configuration.

    Specifies which MCP server tools the agent can use for data retrieval.
    """

    server: str = Field(description="MCP server name (e.g., 'collection', 'plantation')")
    tools: list[str] = Field(description="List of tools to use from this server (e.g., ['get_document', 'get_farmer'])")


class ErrorHandlingConfig(BaseModel):
    """Error handling and retry configuration.

    Configures retry behavior and failure handling for agent execution.
    """

    max_attempts: int = Field(
        default=3,
        ge=1,
        le=10,
        description="Maximum retry attempts",
    )
    backoff_ms: list[int] = Field(
        default_factory=lambda: [100, 500, 2000],
        description="Backoff delays in milliseconds between retries",
    )
    on_failure: Literal["publish_error_event", "dead_letter", "graceful_fallback"] = Field(
        default="publish_error_event",
        description="Action on final failure",
    )
    dead_letter_topic: str | None = Field(
        default=None,
        description="Dead letter topic for failed messages (if on_failure='dead_letter')",
    )


class StateConfig(BaseModel):
    """Conversation state management configuration.

    Used only by Conversational agents for multi-turn dialogue.
    """

    max_turns: int = Field(
        default=5,
        ge=1,
        le=50,
        description="Maximum turns in a conversation",
    )
    session_ttl_minutes: int = Field(
        default=30,
        ge=1,
        le=1440,
        description="Session timeout in minutes",
    )
    checkpoint_backend: Literal["mongodb"] = Field(
        default="mongodb",
        description="Backend for state persistence",
    )
    context_window: int = Field(
        default=3,
        ge=1,
        le=20,
        description="Number of previous turns to include in prompt",
    )


class AgentConfigMetadata(BaseModel):
    """Agent configuration metadata for tracking changes.

    Stores authorship, timestamps, and version control info.
    """

    author: str = Field(description="Author username or system identifier")
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC),
        description="Creation timestamp",
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC),
        description="Last update timestamp",
    )
    git_commit: str | None = Field(
        default=None,
        description="Source commit SHA for traceability",
    )


# =============================================================================
# TIERED-VISION SPECIFIC MODELS
# =============================================================================


class TieredVisionLLMConfig(BaseModel):
    """Two-tier LLM configuration for vision processing.

    Cost-optimized: Haiku for fast screening, Sonnet for deep analysis.
    """

    screen: LLMConfig = Field(description="Tier 1: Fast screening model (typically Haiku)")
    diagnose: LLMConfig = Field(description="Tier 2: Deep analysis model (typically Sonnet)")


class TieredVisionRoutingConfig(BaseModel):
    """Routing configuration for tiered vision processing.

    Determines when to escalate from Tier 1 (screen) to Tier 2 (diagnose).
    """

    screen_threshold: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Escalate to Tier 2 if screen confidence < threshold",
    )
    healthy_skip_threshold: float = Field(
        default=0.85,
        ge=0.0,
        le=1.0,
        description="Skip Tier 2 for 'healthy' classification above this confidence",
    )
    obvious_skip_threshold: float = Field(
        default=0.75,
        ge=0.0,
        le=1.0,
        description="Skip Tier 2 for 'obvious_issue' classification above this confidence",
    )


# =============================================================================
# BASE CLASS (common fields)
# =============================================================================


class AgentConfigBase(BaseModel):
    """Base configuration shared by all agent types.

    Contains common fields that all agent configurations must have.
    Type-specific models inherit from this and add their specialized fields.
    """

    id: str = Field(description="Unique document ID (format: {agent_id}:{version})")
    agent_id: str = Field(description="Logical agent identifier (e.g., 'disease-diagnosis')")
    version: str = Field(description="Semantic version string (e.g., '1.0.0')")
    status: AgentConfigStatus = Field(
        default=AgentConfigStatus.DRAFT,
        description="Agent configuration lifecycle status",
    )
    description: str = Field(description="Human-readable description of what this agent does")
    input: InputConfig = Field(description="Input event and schema contract")
    output: OutputConfig = Field(description="Output event and schema contract")
    llm: LLMConfig = Field(description="LLM configuration for agent execution")
    mcp_sources: list[MCPSourceConfig] = Field(
        default_factory=list,
        description="MCP servers and tools this agent can use",
    )
    error_handling: ErrorHandlingConfig = Field(
        default_factory=ErrorHandlingConfig,
        description="Error handling and retry configuration",
    )
    metadata: AgentConfigMetadata = Field(
        description="Agent configuration metadata",
    )


# =============================================================================
# TYPE-SPECIFIC MODELS
# =============================================================================


class ExtractorConfig(AgentConfigBase):
    """Extractor agent: structured data from unstructured input.

    Uses single LLM call for fast, deterministic extraction.
    No RAG required - works directly on input data.
    """

    type: Literal["extractor"] = Field(
        default="extractor",
        description="Agent type discriminator",
    )
    extraction_schema: dict[str, Any] = Field(description="Schema defining fields to extract and validation rules")
    normalization_rules: list[dict[str, Any]] | None = Field(
        default=None,
        description="Optional rules for value normalization (e.g., uppercase, prefix)",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "qc-event-extractor:1.0.0",
                "agent_id": "qc-event-extractor",
                "version": "1.0.0",
                "type": "extractor",
                "status": "active",
                "description": "Extracts structured data from QC analyzer payloads",
                "input": {
                    "event": "collection.document.received",
                    "schema": {"required": ["doc_id"]},
                },
                "output": {
                    "event": "ai.extraction.complete",
                    "schema": {"fields": ["farmer_id", "grade", "quality_score"]},
                },
                "llm": {
                    "model": "anthropic/claude-3-haiku",
                    "temperature": 0.1,
                    "max_tokens": 500,
                },
                "extraction_schema": {"required_fields": ["farmer_id", "grade"]},
            }
        }
    }


class ExplorerConfig(AgentConfigBase):
    """Explorer agent: analyze, diagnose, find patterns.

    Uses RAG for knowledge retrieval to inform analysis.
    Single LLM capable model for complex reasoning.
    """

    type: Literal["explorer"] = Field(
        default="explorer",
        description="Agent type discriminator",
    )
    rag: RAGConfig = Field(description="RAG configuration for knowledge retrieval")

    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "disease-diagnosis:1.0.0",
                "agent_id": "disease-diagnosis",
                "version": "1.0.0",
                "type": "explorer",
                "status": "active",
                "description": "Analyzes quality issues and produces diagnosis",
                "llm": {
                    "model": "anthropic/claude-3-5-sonnet",
                    "temperature": 0.3,
                    "max_tokens": 2000,
                },
                "rag": {
                    "enabled": True,
                    "knowledge_domains": ["plant_diseases", "tea_cultivation"],
                    "top_k": 5,
                },
            }
        }
    }


class GeneratorConfig(AgentConfigBase):
    """Generator agent: create content (plans, reports, messages).

    Uses RAG for best practices and context.
    Supports multiple output formats (json, markdown, text).
    """

    type: Literal["generator"] = Field(
        default="generator",
        description="Agent type discriminator",
    )
    rag: RAGConfig = Field(description="RAG configuration for knowledge retrieval")
    output_format: Literal["json", "markdown", "text"] = Field(
        default="markdown",
        description="Output content format",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "weekly-action-plan:1.0.0",
                "agent_id": "weekly-action-plan",
                "version": "1.0.0",
                "type": "generator",
                "status": "active",
                "description": "Generates personalized weekly action plans for farmers",
                "llm": {
                    "model": "anthropic/claude-3-5-sonnet",
                    "temperature": 0.5,
                    "max_tokens": 3000,
                },
                "rag": {
                    "enabled": True,
                    "knowledge_domains": ["tea_cultivation", "regional_context"],
                },
                "output_format": "markdown",
            }
        }
    }


class ConversationalConfig(AgentConfigBase):
    """Conversational agent: multi-turn dialogue with persona.

    Uses two LLMs: fast model for intent, capable model for response.
    Includes RAG for knowledge grounding and state management.
    """

    type: Literal["conversational"] = Field(
        default="conversational",
        description="Agent type discriminator",
    )
    rag: RAGConfig = Field(description="RAG configuration for knowledge grounding")
    state: StateConfig = Field(
        default_factory=StateConfig,
        description="Conversation state management configuration",
    )
    intent_model: str = Field(
        default="anthropic/claude-3-haiku",
        description="Fast model for intent classification",
    )
    response_model: str = Field(
        default="anthropic/claude-3-5-sonnet",
        description="Capable model for generating responses",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "dialogue-responder:1.0.0",
                "agent_id": "dialogue-responder",
                "version": "1.0.0",
                "type": "conversational",
                "status": "active",
                "description": "Handles multi-turn dialogue with farmers via voice/WhatsApp",
                "llm": {"model": "anthropic/claude-3-5-sonnet", "temperature": 0.4},
                "rag": {
                    "enabled": True,
                    "knowledge_domains": ["tea_cultivation", "common_questions"],
                },
                "state": {"max_turns": 5, "session_ttl_minutes": 30},
                "intent_model": "anthropic/claude-3-haiku",
                "response_model": "anthropic/claude-3-5-sonnet",
            }
        }
    }


class TieredVisionConfig(AgentConfigBase):
    """Tiered-Vision agent: cost-optimized image analysis.

    Two-tier processing:
    - Tier 1 (screen): Fast Haiku classification on thumbnail
    - Tier 2 (diagnose): Deep Sonnet analysis on full image (if needed)

    Routing config determines when to escalate to Tier 2.
    """

    type: Literal["tiered-vision"] = Field(
        default="tiered-vision",
        description="Agent type discriminator",
    )
    # Override base llm to be optional (replaced by tiered_llm)
    llm: LLMConfig | None = Field(
        default=None,
        description="Not used - tiered_llm replaces this for tiered-vision agents",
    )
    rag: RAGConfig = Field(description="RAG configuration (used in Tier 2 only)")
    tiered_llm: TieredVisionLLMConfig = Field(description="Two-tier LLM configuration (screen + diagnose)")
    routing: TieredVisionRoutingConfig = Field(
        default_factory=TieredVisionRoutingConfig,
        description="Routing thresholds for tier escalation",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "leaf-quality-analyzer:1.0.0",
                "agent_id": "leaf-quality-analyzer",
                "version": "1.0.0",
                "type": "tiered-vision",
                "status": "active",
                "description": "Cost-optimized image analysis for tea leaf quality",
                "llm": None,
                "rag": {
                    "enabled": True,
                    "knowledge_domains": ["plant_diseases", "visual_symptoms"],
                },
                "tiered_llm": {
                    "screen": {
                        "model": "anthropic/claude-3-haiku",
                        "temperature": 0.1,
                        "max_tokens": 200,
                    },
                    "diagnose": {
                        "model": "anthropic/claude-3-5-sonnet",
                        "temperature": 0.3,
                        "max_tokens": 2000,
                    },
                },
                "routing": {
                    "screen_threshold": 0.7,
                    "healthy_skip_threshold": 0.85,
                },
            }
        }
    }


# =============================================================================
# DISCRIMINATED UNION (automatic type detection)
# =============================================================================

AgentConfig = Annotated[
    ExtractorConfig | ExplorerConfig | GeneratorConfig | ConversationalConfig | TieredVisionConfig,
    Field(discriminator="type"),
]
"""Discriminated union of all agent configuration types.

Usage:
    from pydantic import TypeAdapter

    adapter = TypeAdapter(AgentConfig)
    config = adapter.validate_python({"type": "explorer", ...})  # Returns ExplorerConfig

Pydantic automatically selects the correct type based on the "type" field.
"""

File: web/platform-admin/src/types/agent-config.ts
/**
 * AI Agent Configuration TypeScript Interfaces
 *
 * Type definitions for AI Agent & Prompt Viewer matching BFF API schemas.
 * Mirrors fp_common/models/agent_config_summary.py Pydantic models.
 *
 * Story 9.12c: AI Agent & Prompt Viewer UI
 */

// ============================================================================
// Agent Type and Status Types
// ============================================================================

/** Agent type (maps to AI Model agent types) */
export type AgentType = 'extractor' | 'explorer' | 'generator' | 'conversational' | 'tiered-vision';

/** Agent configuration status */
export type AgentStatus = 'draft' | 'staged' | 'active' | 'archived';

// ============================================================================
// LLM Configuration Types (parsed from config_json)
// ============================================================================

/** Retry configuration for LLM calls */
export interface RetryConfig {
  max_retries?: number;
  backoff?: string;
  timeout_seconds?: number;
}

/** LLM configuration block */
export interface LlmConfig {
  model: string;
  temperature: number;
  max_tokens: number;
  top_p?: number;
  response_format?: string;
  retry?: RetryConfig;
}

// ============================================================================
// RAG Configuration Types (parsed from config_json)
// ============================================================================

/** RAG configuration block */
export interface RagConfig {
  enabled: boolean;
  knowledge_domains?: string[];
  domains?: string[]; // Alias for knowledge_domains
  top_k?: number;
  min_similarity?: number;
  score_threshold?: number; // Alias for min_similarity
  namespace?: string;
  include_metadata?: boolean;
  rerank_enabled?: boolean;
  reranking_model?: string | null;
}

// ============================================================================
// Contract Configuration Types (parsed from config_json)
// ============================================================================

/** Schema definition for contracts */
export interface SchemaDefinition {
  type: string;
  required?: string[];
  properties?: Record<string, unknown>;
}

/** Event configuration */
export interface EventDefinition {
  event: string;
  schema: SchemaDefinition;
}

/** Input contract configuration */
export interface InputContract {
  event?: string;
  schema?: SchemaDefinition;
}

/** Output contract configuration */
export interface OutputContract {
  event?: string;
  schema?: SchemaDefinition;
}

/** Extraction schema for extractor agents */
export interface ExtractionSchema {
  required_fields: string[];
  optional_fields?: string[];
  field_types?: Record<string, string>;
}

/** Error handling configuration */
export interface ErrorHandlingConfig {
  max_attempts?: number;
  backoff_ms?: number[];
  on_failure?: string;
  dead_letter_topic?: string | null;
}

/** Metadata block */
export interface AgentMetadata {
  author?: string;
  created_at?: string;
  updated_at?: string;
  changelog?: string;
  git_commit?: string | null;
}

// ============================================================================
// Complete Agent Configuration (parsed from config_json)
// ============================================================================

/** Complete agent configuration parsed from config_json */
export interface AgentConfig {
  agent_id: string;
  version: string;
  type: AgentType;
  status: AgentStatus;
  description: string;

  // Event configuration
  input?: InputContract;
  output?: OutputContract;

  // LLM configuration
  llm: LlmConfig;

  // RAG configuration (optional, mainly for explorer agents)
  rag?: RagConfig;

  // Extraction schema (for extractor agents)
  extraction_schema?: ExtractionSchema;

  // MCP sources
  mcp_sources?: string[];

  // Error handling
  error_handling?: ErrorHandlingConfig;

  // Metadata
  metadata?: AgentMetadata;
}

// ============================================================================
// Prompt Types
// ============================================================================

/** Prompt content structure */
export interface PromptContent {
  system_prompt: string;
  template: string;
  output_schema?: Record<string, unknown>;
  few_shot_examples?: Array<{
    input: unknown;
    output: unknown;
  }>;
}

/** Prompt A/B test configuration */
export interface PromptAbTestConfig {
  enabled: boolean;
  traffic_percentage?: number;
  test_id?: string | null;
}

/** Prompt metadata */
export interface PromptMetadata {
  author?: string;
  created_at?: string;
  updated_at?: string;
  changelog?: string;
  git_commit?: string | null;
}

// ============================================================================
// API Response Types
// ============================================================================

/** Prompt summary for list views */
export interface PromptSummary {
  id: string;
  prompt_id: string;
  agent_id: string;
  version: string;
  status: string;
  author: string;
  updated_at: string | null;
}

/** Full prompt detail (for expanded view) */
export interface PromptDetail extends PromptSummary {
  content?: PromptContent;
  metadata?: PromptMetadata;
  ab_test?: PromptAbTestConfig;
}

/** Agent configuration summary for list views */
export interface AgentConfigSummary {
  agent_id: string;
  version: string;
  agent_type: AgentType;
  status: AgentStatus;
  description: string;
  model: string;
  prompt_count: number;
  updated_at: string | null;
}

/** Agent configuration detail response from BFF */
export interface AgentConfigDetail extends AgentConfigSummary {
  config_json: string; // JSON string - must be parsed client-side
  prompts: PromptSummary[];
  created_at: string | null;
}

/** Pagination metadata */
export interface AgentConfigPaginationMeta {
  total_count: number;
  page_size: number;
  next_page_token: string | null;
}

/** Agent configuration list API response */
export interface AgentConfigListResponse {
  data: AgentConfigSummary[];
  pagination: AgentConfigPaginationMeta;
}

/** Agent configuration list query parameters */
export interface AgentConfigListParams {
  page_size?: number;
  page_token?: string;
  agent_type?: AgentType;
  status?: AgentStatus;
}

/** Prompt list response from BFF */
export interface PromptListResponse {
  data: PromptSummary[];
  total_count: number;
}

// ============================================================================
// Helper Functions
// ============================================================================

/** Get display label for agent type */
export function getAgentTypeLabel(type: AgentType): string {
  const labels: Record<AgentType, string> = {
    extractor: 'Extractor',
    explorer: 'Explorer',
    generator: 'Generator',
    conversational: 'Conversational',
    'tiered-vision': 'Tiered Vision',
  };
  return labels[type] || type;
}

/** Get chip color for agent type */
export function getAgentTypeColor(type: AgentType): 'info' | 'warning' | 'success' | 'secondary' | 'primary' {
  const colors: Record<AgentType, 'info' | 'warning' | 'success' | 'secondary' | 'primary'> = {
    extractor: 'info',
    explorer: 'warning',
    generator: 'success',
    conversational: 'secondary',
    'tiered-vision': 'primary',
  };
  return colors[type] || 'info';
}

/** Get display label for status */
export function getStatusLabel(status: AgentStatus | string): string {
  const labels: Record<string, string> = {
    draft: 'Draft',
    staged: 'Staged',
    active: 'Active',
    archived: 'Archived',
  };
  return labels[status] || status;
}

/** Get chip color for status */
export function getStatusColor(status: AgentStatus | string): 'success' | 'warning' | 'default' | 'info' {
  const colors: Record<string, 'success' | 'warning' | 'default' | 'info'> = {
    active: 'success',
    staged: 'warning',
    archived: 'default',
    draft: 'info',
  };
  return colors[status] || 'default';
}

/** Parse config_json string into typed AgentConfig object */
export function parseConfigJson(configJson: string): AgentConfig | null {
  try {
    return JSON.parse(configJson) as AgentConfig;
  } catch {
    return null;
  }
}

/** Get RAG domains from RAG config, handling both field names */
export function getRagDomains(ragConfig?: RagConfig): string[] {
  if (!ragConfig) return [];
  return ragConfig.knowledge_domains || ragConfig.domains || [];
}

/** Check if RAG is enabled */
export function isRagEnabled(ragConfig?: RagConfig): boolean {
  return ragConfig?.enabled ?? false;
}

/** Format date string for display */
export function formatDate(dateString: string | null | undefined): string {
  if (!dateString) return 'N/A';
  try {
    const date = new Date(dateString);
    return date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
  } catch {
    return dateString;
  }
}
